src/private/MDMAnimationRegistrar.m

```diff
@@ -0,0 +1,102 @@
+/*
+ Copyright 2017-present The Material Motion Authors. All Rights Reserved.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+#import "MDMAnimationRegistrar.h"
+
+#import "MDMRegisteredAnimation.h"
+
+@implementation MDMAnimationRegistrar {
+  NSMapTable *_layersToRegisteredAnimation;  // Map of [CALayer:MDMRegisteredAnimation]
+}
+
+- (instancetype)init {
+  self = [super init];
+  if (self) {
+    _layersToRegisteredAnimation = [NSMapTable mapTableWithKeyOptions:NSPointerFunctionsWeakMemory
+                                                      valueOptions:NSPointerFunctionsStrongMemory];
+  }
+  return self;
+}
+
+#pragma mark - Private
+
+- (void)forEachAnimation:(void (^)(CALayer *, CABasicAnimation *, NSString *))work {
+  // Copy the registered animations before iteration in case further modifications happen to the
```

Comment:
Looking at this comment, are you worried that the dictionary can be mutated while you are iterating over it?  Are the `work` blocks the source of mutation?

Reply:
Yep - consider if we remove an animation, its associated completion block might get fired and a new animation might be registered as a result.

Reply:
Got it, thanks!